File: Android/scenes/show/src/main/java/io/agora/scene/show/service/rtmsync/RoomPresenceService.kt
package io.agora.scene.show.service.rtmsync

import androidx.annotation.IntDef
import io.agora.rtmsyncmanager.service.rtm.AUIRtmManager
import io.agora.rtmsyncmanager.service.rtm.AUIRtmUserLeaveReason
import io.agora.rtmsyncmanager.service.rtm.AUIRtmUserRespObserver
import io.agora.rtmsyncmanager.utils.AUILogger
import io.agora.rtmsyncmanager.utils.GsonTools
import io.agora.rtmsyncmanager.utils.ObservableHelper
import io.agora.rtmsyncmanager.utils.ThreadManager

class RoomPresenceService(
    private val rtmManager: AUIRtmManager,
    private val channelName: String
) {
    private val tag = "RoomPresenceService($channelName)"

    private val observerHelper = ObservableHelper<RoomPresenceSubscriber>()

    private val roomPresenceInfoList = mutableListOf<RoomPresenceInfo>()

    private val userRespObserver = object : AUIRtmUserRespObserver {

        override fun onUserSnapshotRecv(
            channelName: String,
            userId: String,
            userList: List<Map<String, Any>>
        ) {
            if (channelName != this@RoomPresenceService.channelName) {
                return
            }
            userList.forEach {
                val info =
                    GsonTools.toBeanSafely(it, RoomPresenceInfo::class.java) ?: return@forEach
                val index = roomPresenceInfoList.indexOfFirst { it.roomId == info.roomId }
                if (index == -1) {
                    roomPresenceInfoList.add(info)
                } else {
                    roomPresenceInfoList[index] = info
                }
            }
            AUILogger.logger().d(tag, "onUserSnapshotRecv ${roomPresenceInfoList.size}")
            observerHelper.notifyEventHandlers {
                it.onSnapshot?.invoke(ArrayList(roomPresenceInfoList))
            }
        }

        override fun onUserDidJoined(
            channelName: String,
            userId: String,
            userInfo: Map<String, Any>
        ) {
            if (channelName != this@RoomPresenceService.channelName) {
                return
            }
            val info = GsonTools.toBeanSafely(userInfo, RoomPresenceInfo::class.java) ?: return
            val index = roomPresenceInfoList.indexOfFirst { it.roomId == info.roomId }
            if (index == -1) {
                roomPresenceInfoList.add(info)
            } else {
                roomPresenceInfoList[index] = info
            }
            AUILogger.logger().d(tag, "onUserDidJoined $info")
            observerHelper.notifyEventHandlers {
                it.onUpdate?.invoke(info)
            }
        }

        override fun onUserDidLeaved(
            channelName: String,
            userId: String,
            userInfo: Map<String, Any>,
            reason: AUIRtmUserLeaveReason
        ) {
            if (channelName != this@RoomPresenceService.channelName) {
                return
            }
            val info = roomPresenceInfoList.findLast { it.ownerId == userId }
            AUILogger.logger().d(tag, "onUserDidLeaved $info")
            if (info != null) {
                roomPresenceInfoList.removeIf { it.ownerId == userId }
                observerHelper.notifyEventHandlers {
                    it.onDelete?.invoke(info)
                }
            }
        }

        override fun onUserDidUpdated(
            channelName: String,
            userId: String,
            userInfo: Map<String, Any>
        ) {
            if (channelName != this@RoomPresenceService.channelName) {
                return
            }
            val info = GsonTools.toBeanSafely(userInfo, RoomPresenceInfo::class.java) ?: return
            val index = roomPresenceInfoList.indexOfFirst { it.roomId == info.roomId }
            if (index == -1) {
                roomPresenceInfoList.add(info)
            } else {
                roomPresenceInfoList[index] = info
            }
            AUILogger.logger().d(tag, "onUserDidUpdated $info")
            observerHelper.notifyEventHandlers {
                it.onUpdate?.invoke(info)
            }
        }

    }

    fun login(complete: () -> Unit) {
        AUILogger.logger().d(tag, "login")
        rtmManager.subscribeUser(userRespObserver)
        rtmManager.subscribe(channelName) { ex ->
            AUILogger.logger().d(tag, "login complete: $ex")
            if (ex != null) {
                observerHelper.notifyEventHandlers {
                    it.onError?.invoke(RuntimeException(ex))
                }
            } else {
                complete.invoke()
            }
        }
    }

    fun logout() {
        AUILogger.logger().d(tag, "logout")
        observerHelper.unSubscribeAll()
        rtmManager.unSubscribe(channelName)
        rtmManager.unsubscribeUser(userRespObserver)
    }

    fun setup(
        info: RoomPresenceInfo,
        success: (() -> Unit)? = null,
        error: ((Exception) -> Unit)? = null
    ) {
        AUILogger.logger().d(tag, "setup")
        setRoomPresenceInfo(info,
            {
                AUILogger.logger().d(tag, "setup success")
                success?.invoke()
            },
            {
                AUILogger.logger().d(tag, "setup failed : $it")
                error?.invoke(it)
            })
    }

    fun subscribe(subscriber: RoomPresenceSubscriber) {
        AUILogger.logger().d(tag, "subscribe $subscriber")
        observerHelper.subscribeEvent(subscriber)
    }

    fun unSubscribe(subscriber: RoomPresenceSubscriber) {
        AUILogger.logger().d(tag, "unSubscribe $subscriber")
        observerHelper.unSubscribeEvent(subscriber)
    }

    private fun setRoomPresenceInfo(
        info: RoomPresenceInfo,
        success: (() -> Unit)? = null,
        error: ((Exception) -> Unit)? = null
    ) {
        AUILogger.logger().d(tag, "setRoomPresenceInfo $info")
        rtmManager.setPresenceState(
            channelName,
            attr = GsonTools.beanToMap(info)
        ) { ex ->
            AUILogger.logger().d(tag, "setRoomPresenceInfo complete: $ex")
            if (ex != null) {
                ThreadManager.getInstance().runOnMainThread {
                    error?.invoke(RuntimeException(ex))
                }
                return@setPresenceState
            }
            ThreadManager.getInstance().runOnMainThread {
                val index = roomPresenceInfoList.indexOfFirst { it.roomId == info.roomId }
                if (index == -1) {
                    roomPresenceInfoList.add(info)
                } else {
                    roomPresenceInfoList[index] = info
                }
                observerHelper.notifyEventHandlers {
                    it.onUpdate?.invoke(info)
                }
                success?.invoke()
            }
        }
    }

    fun updateRoomPresenceInfo(
        roomId: String,
        @RoomPresenceStatus status: Int,
        interactorId: String = "",
        interactorName: String = "",
        success: (() -> Unit)? = null,
        error: ((Exception) -> Unit)? = null
    ) {
        AUILogger.logger()
            .d(tag, "updateRoomPresenceInfo $roomId $status $interactorId $interactorName")
        val interactionInfo = getRoomPresenceInfo(roomId)
        if (interactionInfo == null) {
            AUILogger.logger().d(tag, "updateRoomPresenceInfo failed: RoomInteractionInfo not found")
            error?.invoke(RuntimeException("RoomInteractionInfo not found"))
            return
        }
        setRoomPresenceInfo(
            interactionInfo.copy(
                status = status,
                interactorId = interactorId,
                interactorName = interactorName
            ),
            {
                AUILogger.logger().d(tag, "updateRoomPresenceInfo success")
                success?.invoke()
            },
            {
                AUILogger.logger().d(tag, "updateRoomPresenceInfo failed: $it")
                error?.invoke(it)
            }
        )
    }

    fun getAllRoomPresenceInfo(
        success: (List<RoomPresenceInfo>) -> Unit,
        error: ((Exception) -> Unit)? = null
    ) {
        AUILogger.logger().d(tag, "getAllRoomPresenceInfo")
        rtmManager.whoNow(channelName) { ex, userList ->
            AUILogger.logger().d(tag, "getAllRoomPresenceInfo complete: $ex, userListSize=${userList?.size}")
            if (ex != null) {
                error?.invoke(RuntimeException(ex))
                return@whoNow
            }
            val list = mutableListOf<RoomPresenceInfo>()
            userList?.forEach {
                if (it.size <= 1) {
                    return@forEach
                }
                val info =
                    GsonTools.toBeanSafely(it, RoomPresenceInfo::class.java) ?: return@forEach
                list.add(info)
            }
            ThreadManager.getInstance().runOnMainThread {
                success.invoke(list)
            }
        }
    }

    fun getRoomPresenceInfo(roomId: String): RoomPresenceInfo? {
        val firstOrNull = roomPresenceInfoList.firstOrNull { it.roomId == roomId }
        AUILogger.logger().d(tag, "getRoomPresenceInfo $roomId $firstOrNull")
        return firstOrNull
    }

    fun getRoomPresenceInfoByOwnerId(ownerId: String): RoomPresenceInfo? {
        val firstOrNull = roomPresenceInfoList.lastOrNull { it.ownerId == ownerId }
        AUILogger.logger().d(tag, "getRoomPresenceInfoByOwnerId $ownerId $firstOrNull")
        return firstOrNull
    }
}

@IntDef(
    RoomPresenceStatus.IDLE,
    RoomPresenceStatus.INTERACTING_LINKING,
    RoomPresenceStatus.INTERACTING_PK
)
@Retention(AnnotationRetention.RUNTIME)
annotation class RoomPresenceStatus {
    companion object {
        const val IDLE = 0
        const val INTERACTING_LINKING = 1
        const val INTERACTING_PK = 2
    }
}

data class RoomPresenceInfo constructor(
    val roomId: String, // 唯一房间ID
    val roomName: String, // 房间名
    val ownerId: String, // 房主用户ID
    val ownerName: String, // 房主名
    val ownerAvatar: String, // 房主头像
    @RoomPresenceStatus val status: Int = RoomPresenceStatus.IDLE,
    val interactorId: String = "", // 互动者ID
    val interactorName: String = ""// 互动者名
)

data class RoomPresenceSubscriber(
    val onSnapshot: ((List<RoomPresenceInfo>) -> Unit)? = null,
    val onUpdate: ((RoomPresenceInfo) -> Unit)? = null,
    val onDelete: ((RoomPresenceInfo) -> Unit)? = null,
    val onError: ((Exception) -> Unit)? = null
)

File: iOS/AgoraEntScenarios/Common/API/RTMSyncManager/RTMSyncManager/Classes/Extension/ApplyAndInvitation/RoomPresenceService.swift
//
//  RoomPresenceService.swift
//  RTMSyncManager
//
//  Created by wushengtao on 2024/6/7.
//

import Foundation

@objcMembers public class RoomPresenceInfo: NSObject, Codable {
    public var roomId: String = "" // 唯一房间ID
    public var roomName: String = "" // 房间名
    public var ownerId: String = "" // 房主用户ID
    public var ownerName: String = "" // 房主名
    public var ownerAvatar: String = "" // 房主头像
    public var status: InteractionType = .idle
    public var interactorId: String = "" // 互动者ID
    public var interactorName: String = ""  // 互动者名
    
    private enum CodingKeys: String, CodingKey {
        case roomId
        case roomName
        case ownerId
        case ownerName
        case ownerAvatar
        case status
        case interactorId
        case interactorName
    }
        
//    required public init(from decoder: Decoder) throws {
//        let container = try decoder.singleValueContainer()
//        let statusValue = try container.decode(String.self)
//        
//        if let statusInt = Int(statusValue), let statusEnum = RoomPresenceStatus(rawValue: statusInt) {
//            self.status = statusEnum
//        }
//        
//        super.init()
//    }
}

private func convertMap(_ map: [String: Any]) -> [String: Any] {
    var convertMap = map
    convertMap["status"] = Int("\(map["status"] as? String ?? "")") ?? 0
    return convertMap
}

@objc public protocol RoomPresenceProtocol: NSObjectProtocol {
    func onUserSnapshot(channelName: String, userList: [RoomPresenceInfo])
    func onUserUpdate(channelName: String, user: RoomPresenceInfo)
    func onUserDelete(channelName: String, user: RoomPresenceInfo)
    func onUserError(channelName: String, error: NSError)
}

public class RoomPresenceService: NSObject {
    private var rtmManager: AUIRtmManager
    private var channelName: String
    private var respDelegates = NSHashTable<RoomPresenceProtocol>.weakObjects()
    private(set) var userList: [RoomPresenceInfo] = []
    
    required init(channelName: String, rtmManager: AUIRtmManager) {
        self.channelName = channelName
        self.rtmManager = rtmManager
        super.init()
    }
    
    public func subscribeChannel() {
        rtmManager.subscribeUser(channelName: channelName, delegate: self)
        rtmManager.subscribe(channelName: channelName) { err in
            
        }
    }
    
    public func unsubscribeChannel() {
        rtmManager.unSubscribe(channelName: channelName)
    }
}

//MARK: public
extension RoomPresenceService {
    public func subscribe(delegate: RoomPresenceProtocol) {
        if respDelegates.contains(delegate) {
            return
        }
        
        respDelegates.add(delegate)
    }
    
    public func unsubscribe(delegate: RoomPresenceProtocol) {
        respDelegates.remove(delegate)
    }
    
    public func setRoomPresenceInfo(user: RoomPresenceInfo, completion: ((NSError?) -> ())?) {
        aui_info("setRoomPresenceInfo ownerId: \(user.ownerId) ownerName: \(user.ownerName) roomId: \(user.roomId)", tag: "RoomPresenceService")
        guard let attr = encodeModel(user) else {
            completion?(NSError(domain: "RoomPresenceService",
                                code: 0,
                                userInfo: ["msg": "encodeModel fail"]))
            return
        }
        rtmManager.setPresenceState(channelName: channelName, attr: attr) { err in
            completion?(err)
        }
    }
    
    public func updateRoomPresenceInfo(roomId: String,
                                       status: InteractionType,
                                       interactorId: String,
                                       interactorName: String,
                                       completion: ((NSError?) -> ())?) {
        aui_info("updateRoomPresenceInfo roomId: \(roomId) status: \(status.rawValue) interactorId: \(interactorId) interactorName: \(interactorName)", tag: "RoomPresenceService")
        guard let user = getRoomPresenceInfo(roomId: roomId) else {
          completion?(NSError(domain: "RoomPresenceService",
                              code: 0,
                              userInfo: ["msg": "RoomInteractionInfo not found"]))
            return
        }
        user.status = status
        user.interactorId = interactorId
        user.interactorName = interactorName
        setRoomPresenceInfo(user: user, completion: completion)
    }
    
    public func getRoomPresenceInfo(roomId: String) -> RoomPresenceInfo? {
        let user = userList.first { $0.roomId == roomId }
        return user
    }
    
    public func getRoomPresenceInfoByOwnerId(ownerId: String) -> RoomPresenceInfo? {
        let user = userList.first { $0.ownerId == ownerId }
        return user
    }
    
    public func getAllRoomPresenceInfo(completion: @escaping (NSError?, [RoomPresenceInfo]?) -> ()) {
        rtmManager.whoNow(channelName: channelName) {[weak self] err, userList in
            guard let self = self else {return}
            if let err = err {
                completion(err, nil)
                return
            }
            let convertUserList = userList?.map { convertMap($0)} ?? []
            let userList: [RoomPresenceInfo] = decodeModelArray(convertUserList) ?? []
            self.userList = userList
            completion(nil, userList)
        }
    }
}

extension RoomPresenceService: AUIRtmUserProxyDelegate {
    
    
    public func onCurrentUserJoined(channelName: String) {
        
    }
    
    public func onUserSnapshotRecv(channelName: String, userId: String, userList: [[String : Any]]) {
        aui_info("onUserSnapshotRecv user count: \(userList.count)", tag: "RoomPresenceService")
        let convertUserList = userList.map { convertMap($0) }
        let userList: [RoomPresenceInfo] = decodeModelArray(convertUserList) ?? []
        self.userList = userList
        respDelegates.allObjects.forEach { delegate in
            delegate.onUserSnapshot(channelName: channelName, userList: userList)
        }
    }
    
    public func onUserDidJoined(channelName: String, userId: String, userInfo: [String : Any]) {
//        guard let user: RoomPresenceInfo = decodeModel(userInfo) else {return}
        
//        respDelegates.allObjects.forEach { delegate in
//            delegate.onUserUpdate(user: user)
//        }
    }
    
    public func onUserDidLeaved(channelName: String, userId: String, userInfo: [String : Any], reason: AUIRtmUserLeaveReason) {
        aui_info("onUserDidLeaved userId: \(userId)", tag: "RoomPresenceService")
        guard let user = userList.filter({ $0.ownerId == userId }).first else {return}
        self.userList.removeAll { $0.ownerId == userId }
        respDelegates.allObjects.forEach { delegate in
            delegate.onUserDelete(channelName: channelName, user: user)
        }
    }
    
    public func onUserDidUpdated(channelName: String, userId: String, userInfo: [String : Any]) {
        aui_info("onUserDidUpdated userInfo: \(userInfo)", tag: "RoomPresenceService")
        guard let user: RoomPresenceInfo = decodeModel(convertMap(userInfo)) else {return}
        
        if let index = self.userList.firstIndex(where: { $0.ownerId == userId }) {
            userList[index] = user
        } else {
            userList.append(user)
        }
        
        respDelegates.allObjects.forEach { delegate in
            delegate.onUserUpdate(channelName: channelName, user: user)
        }
    }
}
